Add generics tutorial

expipiplus1 opened the pull request:
Closes https://github.com/shader-slang/slang/issues/5384

csyonghe reviewed (approved):
This is good, but we still need more explanation on the difference between template and generics for people coming from the C++ world with examples that work in C++ but doesn't in Slang, and explain why.

For example,
```
template<T>
float addValue(T v0, T v1) { return v0.x + v1.x; }

void user() {addValue(float3(1,2,3), int2(1,2)); }
```
will work in C++, but the superficial translation won't work in Slang:
```
float addValue<T>(T v0, T v1) { return v0.x + v1.x; }

void user() {addValue(float3(1,2,3), int2(1,2)); }
```
We need to explain why.

Then we need to provide an example of what people often do with HLSL template: abstracting computation over scalar and vector, and show how it is done in Slang (with `IFloat` interface).

expipiplus1 commented:
thanks, I added that section. I simplified the example a little bit to avoid the implicit casting, we should probably make sure to address casting rules somewhere else however.
